lib/carto/oauth_provider/scopes.rb

```
@@ -67,6 +78,70 @@ def initialize(service, description)
         end
       end
 
+      class DatasetsScope < DefaultScope
+
+        READ_PERMISSIONS = ['select'].freeze
+        WRITE_PERMISSIONS = ['insert', 'update', 'delete'].freeze
+
+        PERMISSIONS = {
+          r: READ_PERMISSIONS,
+          w: WRITE_PERMISSIONS,
+          rw: READ_PERMISSIONS + WRITE_PERMISSIONS
+        }.freeze
+
+        DESCRIPTIONS = {
+          r: "Read access to %<table_name>s",
+          w: "Write access to %<table_name>s",
+          rw: "Full access to %<table_name>s"
+        }.freeze
+
+        def initialize(permission, table)
+          super('database', permission, CATEGORY_DATASETS, description(permission.to_sym, table))
+          @grant_key = :tables
+          @permission = permission.to_sym
+          @table = table
+        end
+
+        def description(permission = @permission, table = @table)
+          DESCRIPTIONS[permission] % { table_name: table }
+        end
+
+        def permission
+          PERMISSIONS[@permission]
+        end
+
+        def add_to_api_key_grants(grants, user = nil)
+          ensure_includes_apis(grants, ['maps', 'sql'])
+          database_section = grant_section(grants)
+
+          table_section = {
+            name: @table,
+            permissions: permission,
+            schema: user.database_schema
+          }
+
+          database_section[@grant_key] << table_section
+        end
+
+        def self.is_a?(scope)
+          scope =~ /datasets:(?:rw|w|r):\w+/
+        end
+
+        def self.table(scope)
+          scope.split(':')[-1]
+        end
+
+        def self.valid_scopes(scopes, user)
+          datasets_scopes = scopes.select { |scope| DatasetsScope.is_a?(scope) }.map do |scope|
+            [table(scope), scope]
+          end
+
+          return [] unless datasets_scopes.any?
+          user_tables = user.db_service.tables_effective(user.database_schema)
```

:thinking: Most likely we don't want to validate this all the time. Consider the following:
1. Create dataset A
2. Grant access to A via OAuth
3. Rename A to B
4. Request access to B via OAuth.
  - This will try to upgrade the existing `oauth_app_user` and the validation will fail because `A` no longer exists.

Also, it's quite expensive to check.

Maybe we should only check this when creating an API Key (access token) or when showing the consent screen (fail fast), but not in each validation of the scopes (since a lot of objects validate those).

What do you think?

---

I see.

Actually we are already rising and handling the `RelationDoesNotExistError` on api key creation, so we should be covered if we remove this, but we'll have to update the `oauth_app_user` scopes on `authorize`.

So I guess it's best to fail fast (on consent), let me see what I can do.